### ISSUE TITLE
SSL error when invoking direct method call to IOT device

### ISSUE THREAD

**Issue by its-phil** [Azure/iotedgehubdev#258]
I have the following scenario:
- iotedgehubdev version 0.13.0 running on PC with Windows 10 1809.
- IOT device running Windows IOT core. One module registered.
- Both devices registered with IOT Hub using symmetric keys.

These actions can be performed successfully:
- invoke direct method from IOT Hub to module on iotedgehubdev,
- send message from IOT device to module on iotedgehubdev.

The issue: when invoking a direct method call from the module on iotedgehubdev to IOT device (by calling `moduleClient.InvokeMethodAsync`) I get the following exception:
`System.Security.Authentication.AuthenticationException: The remote certificate is invalid according to the validation procedure`.

The issue does not occur on a real iotedge system.

Am I missing something or is this an actual issue in iotedgehubdev?

**Comment by adashen** [comment]
@its-phil So do you mean you could make an moduleClient.InvokeMethodAsync call from one edge device to another edge device? Both of them installed real edge runtime?

**Comment by its-phil** [comment]
@adashen My IOT device is a dev board running IOT core. If I run the code that I use on iotedgehubdev in an edge module on an iotedge system (Raspberry Pi) I can call moduleClient.InvokeMethodAsync to the IOT device without problems.

**Comment by adashen** [comment]
@its-phil thanks for the information. We will investigate the issue and let you know the updates.

**Comment by blackchoey** [comment]
@its-phil Can you share repro steps for your problem? It would be great if you can also share some sample code here. I wasn't able to repro the issue you mentioned.

**Comment by its-phil** [comment]
@blackchoey thanks for your reply. I'll try to list the steps I performed as detailed as I can remember:

iotedgehubdev is running on a PC, lets call it 'pc'.
Lets call the IOT device 'device' with ID 'deviceIdOfDevice'.

1. In the IoT hub on Azure create an IoT Edge device for 'pc' and an IoT device for 'device'.
2. Set up iotedgehubdev with the command `iotedgehubdev setup -c "<connectionString>"`. The `connectionString` is the one copied from IoT Hub for the 'pc'.
3. Copy the edge-device-ca-root, edge-chain-ca and edge-device-ca certificates to the 'device' and install them.
4. With "Azure IoT Edge Tools for VS 2019" installed:
   1. create an Azure IoT Edge project, 
   2. create an AzureIoTEdgeModule project
   3. assign the module to the to the Azure IoT Edge project.
   4. run the solution from Visual Studio. This produces the following output:

            1>------ Build started: Project: PhilsEdgeHubDev, Configuration: Debug Any CPU ------
            1>DOCKER_HOST:
            1>DOCKER_TLS_VERIFY:
            1>DOCKER_CERT_PATH:
            1>Additional Env Variables:
            1>iotedgehubdev start -d "D:\Projects\PhilsEdgeHubDev\PhilsEdgeHubDev\obj\tmp_deployment_for_local_debug.json"
            1>Stopping edgeHubDev ...
            1>Stopping edgeHubDev ... done
            1>Removing edgeHubDev ...
            1>Removing edgeHubDev ... done
            1>Network azure-iot-edge-dev is external, skipping
            1>Pulling edgeHubDev ...
            1>Pulling edgeHubDev ... pulling from azureiotedge-hub
            1>Pulling edgeHubDev ... digest: sha256:<hash>...
            1>Pulling edgeHubDev ... status: image is up to date for m...
            1>Pulling edgeHubDev ... done
            1>Creating edgeHubDev ...
            1>Creating edgeHubDev ... done
            1>IoT Edge Simulator has been started in solution mode.
            1>Starting to check connectivity for local port 8883...
            1>Successfully connect to local port 8883.
            ========== Build: 1 succeeded, 0 failed, 2 up-to-date, 0 skipped ==========


5. On 'device': connect to hub with `Client = DeviceClient.CreateFromConnectionString(<DeviceConnectionString>, TransportType.Amqp);`. The `DeviceConnectionString` is the one copied from IoT Hub for the 'device'.
6. On 'device': register direct method with `Client.SetMethodHandlerAsync("ConfigureMeasurement", OnConfigureMeasurement, null).Wait();`.
7. In AzureIoTEdgeModule: create ModuleClient instance with

        ITransportSettings[] settings = { new AmqpTransportSettings(TransportType.Amqp_Tcp_Only) };
        var moduleClient = await ModuleClient.CreateFromEnvironmentAsync(settings);

9. On 'device': sending messages to iotedgehubdev works by calling `Client.SendEventAsync(new Message());`.
9. In the AzureIoTEdgeModule project: invoking direct method with `moduleClient.InvokeMethodAsync(deviceIdOfDevice, new MethodRequest("ConfigureMeasurement")).Wait();` fails with the SSL exception mentioned in my initial question.

The last step works when I deploy the solution to an IoT Edge runtime running on a Raspberry Pi.

**Comment by blackchoey** [comment]
@its-phil Thanks for your patient. Seems you're trying to connect your 'device' as downstream device to 'pc'. But I'm still unable to repro your problem under the downstream device scenario: I connected my 'device' as a downstream device to iotedgehubdev, the `InvokeMethodAsync` method call from IoT Edge module to 'device' succeeded.
Can you share whether you appended the `GatewayHostName=xxx` part to your device's connection string in your step 5? This can help me verify whether you're under the downstream device scenario or not.
> 5. On 'device': connect to hub with Client = DeviceClient.CreateFromConnectionString(<DeviceConnectionString>, TransportType.Amqp);. The DeviceConnectionString is the one copied from IoT Hub for the 'device'.

**Comment by its-phil** [comment]
@blackchoey Indeed, I append the GatewayHostName in step 5, so in your terms I am trying to configure the 'device' as downstream to the 'pc'. You mentioned that the InvokeMethodAsync call was successful from 'pc' to 'device'. This tells me that the direct method call should work in general. Did you have both your iotedgehubdev and your device registered with an IoT Hub?

**Comment by its-phil** [comment]
@blackchoey There's an interesting development: when I execute the project in Release mode from Visual Studio everything works just fine, i.e. I can invoke direct methods from iotedgehubdev to the downstream device. The SSL error only occurs in Debug mode.
I can't say that I understand why but maybe it helps you figure out what is going on.

**Comment by blackchoey** [comment]
@its-phil The device used by iotedgehubdev and my physical device is under same IoT Hub. It's interesting that the SSL error only occurs in Debug mode. @SLdragon Can you help take a look?

**Comment by its-phil** [comment]
I think I figured out what goes wrong. Instead of using the Nuget package for Microsoft.Azure.Devices.Client I used the sources so I could debug the code. When invoking the direct method my Edge module tries to send a request to https://localhost/...
As it turns out, on my PC VmWare was serving something on https://localhost. The certificate was issued by VmWare so it is invalid for the Edge module's request. After disabling the VmWare services the direct method calls work as expected. I will monitor the behavior for a while, but for now I think the issue is solved for me.
